### PR TITLE
Solve assignment bugs

### LINKED ISSUE #137: Delete assignment shows incorrect UI
When an assignment is deleted, it always shows 0 student who has completed the assignment.

### LINKED ISSUE #145: Add assignment bug
When a long assignment name (tested with >50M characters) is provided for an assignment, the assignment is saved successfully, but the application crashes, particularly when executing assignment-related commands.

### LINKED ISSUE #146: Mark assignment bug
I marked a student's assignment as done, but after updating the student's ID, the assignment status shows a total of 1 completed, yet the previously marked student is now marked as incomplete.
image
image
image

### LINKED ISSUE #148: Add assignment due date bug
The system allows adding an assignment with an invalid due date, like 2024-02-30 12:00, which is not a valid date in February 2024.
image

### LINKED ISSUE #157: Assignment display bug
image
For assignment detail, it seems like students remain as plural even if there are 0 or 1 student who completed. Perhaps can change this for better user experience.

### LINKED ISSUE #163: Mark Assignment bug
image

First, I addAsg Assignment 1 d/<duedate + time>, 
I also add students as well as shown in the picture.
then I markAsg 0 n/Assignment 1. 

Expected error given: The student index provided is invalid.
Actual error given: Invalid format.

### PR DESCRIPTION
**Pull request by waihongteh**
Closes #163 
Closes #157 
Closes #148 
Closes #137
Closes #145 
Closes #146

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 1]
```diff
@@ -41,6 +42,7 @@ public class AddressBookParser {
      */
     private static final Pattern BASIC_COMMAND_FORMAT = Pattern.compile("(?<commandWord>\\S+)(?<arguments>.*)");
     private static final Logger logger = LogsCenter.getLogger(AddressBookParser.class);
+    private static final int maxCharLimit = 500;
```
Comment (Fui03) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java:45]: Is a 500-character limit too long for an assignment name?

File: src/test/java/seedu/address/model/assignment/AssignmentListTest.java [hunk 2]
```diff
@@ -138,4 +139,18 @@ public void resetData_withValidReadOnlyAddressBook_replacesData() {
         assertEquals(newData, assignments);
     }
 
+    @Test
+    public void reassignStudent_replacesData() {
+        AssignmentList assignments = getTypicalAssignmentList();
+        Student oldStudent = new StudentBuilder().withStudentId("A1234567I").build();
```
Comment (Reallyeasy1) [inline src/test/java/seedu/address/model/assignment/AssignmentListTest.java:145]: Good that you have solved the bugs, would be better to solve the bugs in multiple commits/branches

File: src/test/java/seedu/address/logic/commands/MarkAssignmentCommandTest.java [hunk 3]
```diff
@@ -46,6 +47,22 @@ public void execute_assignmentDoesNotExist_throwsCommandException() {
...
     }
 
+    @Test
+    public void execute_invalidIndex_throwsCommandException() {
+        ReadOnlyAddressBook addressBook = TypicalStudents.getTypicalAddressBook();
+        AssignmentList assignmentList = new AssignmentList();
+        Model model = new ModelManager(addressBook, new UserPrefs(), assignmentList, new TutorialList());
+        MarkAssignmentCommand markCommand1 = new MarkAssignmentCommand(-1, ASSIGNMENT1);
+        MarkAssignmentCommand markCommand2 = new MarkAssignmentCommand(10000, ASSIGNMENT1);
+
```
Comment (Reallyeasy1) [inline src/test/java/seedu/address/logic/commands/MarkAssignmentCommandTest.java:57]: Good job on solving the bugs responsively

File: src/test/java/seedu/address/logic/parser/AddressBookParserTest.java [hunk 4]
```diff
@@ -45,6 +46,18 @@ public class AddressBookParserTest {
 
     private final AddressBookParser parser = new AddressBookParser();
 
+    @Test
+    public void parseCommand_exceedCharLimit_throwsParseException() {
```
Comment (Fui03) [inline src/test/java/seedu/address/logic/parser/AddressBookParserTest.java:50]: I like how you tested the implemented code by using a character length of 10,000.

**Review by Fui03 — COMMENTED** [review]
Great job on your efforts to debug the alpha-bug!

**Pull request merged**